glean-core/src/upload/mod.rs

```diff
@@ -94,98 +123,171 @@ impl PingUploadManager {
         // Verify request was returned
-        assert!(request.is_some());
+        match upload_manager.get_next_ping() {
+            PingUploadTask::Upload(_) => {}
+            _ => panic!("Expected upload manager to return the next request!"),
+        }
     }
 
     #[test]
-    fn test_returns_as_many_ping_request_as_there_are() {
-        let data_dir = tempfile::tempdir().unwrap();
-        let tmpname = data_dir.path().display().to_string();
-
+    fn test_returns_as_many_ping_requests_as_there_are() {
         // Create a new upload_manager
+        let dir = tempfile::tempdir().unwrap();
+        let tmpname = dir.path().display().to_string();
         let upload_manager = PingUploadManager::new(&tmpname);
 
+        // Wait for processing of pending pings directory to finish.
+        while upload_manager.get_next_ping() == PingUploadTask::Wait {
+            thread::sleep(Duration::from_millis(10));
+        }
+
         // Enqueue a ping multiple times
         let n = 10;
         for _ in 0..n {
-            upload_manager.enqueue_ping(UUID, URL, json!({}));
+            upload_manager.enqueue_ping(UUID, URL, json!("{}"));
         }
 
         // Verify a request is returned for each submitted ping
         for _ in 0..n {
-            assert!(upload_manager.get_next_ping().is_some());
+            match upload_manager.get_next_ping() {
+                PingUploadTask::Upload(_) => {}
+                _ => panic!("Expected upload manager to return the next request!"),
+            }
         }
 
         // Verify that after all requests are returned, none are left
-        assert!(upload_manager.get_next_ping().is_none());
+        assert_eq!(upload_manager.get_next_ping(), PingUploadTask::Done);
     }
 
     #[test]
-    fn clearing_the_queue_works_correctly() {
-        let data_dir = tempfile::tempdir().unwrap();
-        let tmpname = data_dir.path().display().to_string();
-
+    fn test_clearing_the_queue_works_correctly() {
         // Create a new upload_manager
+        let dir = tempfile::tempdir().unwrap();
+        let tmpname = dir.path().display().to_string();
         let upload_manager = PingUploadManager::new(&tmpname);
 
+        // Wait for processing of pending pings directory to finish.
+        while upload_manager.get_next_ping() == PingUploadTask::Wait {
+            thread::sleep(Duration::from_millis(10));
+        }
+
+        // Enqueue a ping multiple times
+        for _ in 0..10 {
+            upload_manager.enqueue_ping(UUID, URL, json!("{}"));
+        }
+
+        // Clear the queue
+        upload_manager.clear_ping_queue();
+
+        // Verify there really isn't any ping in the queue
+        assert_eq!(upload_manager.get_next_ping(), PingUploadTask::Done);
+    }
+
+    #[test]
+    fn test_fills_up_queue_successfully_from_disk() {
+        let (mut glean, dir) = new_glean(None);
+
+        // Register a ping for testing
+        let ping_type = PingType::new("test", true, true, vec![]);
```

Comment:
```suggestion
        let ping_type = PingType::new("test", true, /* send_if_empty */ true, vec![]);
```

sometimes it would help having named arguments.
IMO spelling it out this way (or by adding it to the comment before this line) makes it clearer why we can submit the ping multiple times.